server/src/main/java/org/opensearch/index/shard/RemoteStoreRefreshListener.java

```diff
@@ -59,6 +59,13 @@ public RemoteStoreRefreshListener(IndexShard indexShard) {
             .getDelegate()).getDelegate();
         this.primaryTerm = indexShard.getOperationPrimaryTerm();
         localSegmentChecksumMap = new HashMap<>();
+        if (indexShard.shardRouting.primary()) {
+            try {
+                this.remoteDirectory.init();
+            } catch (IOException e) {
+                logger.error("Exception while initialising RemoteSegmentStoreDirectory", e);
+            }
+        }
```

Comment:
Should this logic be sitting inside `indexShard.remoteStore().directory()` instead?

Reply:
It is present in initialization of `RemoteSegmentStoreDirectory`. But when IndexShard is created with remote directory, during failover, only engine of the shard changes. For new primary, this engine would be `InternalEngine` and new instance of RemoteStoreRefreshListener is created but the remote directory instance is still old (at the time of IndexShard creation). So the remote directory instance is not aware of segments that were uploaded by old primary.

Reply:
Let me think more on abstracting out `init()` to prevent misuse. Created issue to track this: https://github.com/opensearch-project/OpenSearch/issues/4410